service/controller/resource/instance/create_terminate_old_vmss.go

```diff
@@ -0,0 +1,51 @@
+package instance
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/giantswarm/microerror"
+
+	"github.com/giantswarm/azure-operator/service/controller/internal/state"
+	"github.com/giantswarm/azure-operator/service/controller/key"
+)
+
+const (
+	legacyVMSSDeploymentName = "worker-vmss-deploy"
+)
+
+func (r *Resource) terminateOldVmssTransition(ctx context.Context, obj interface{}, currentState state.State) (state.State, error) {
+	cr, err := key.ToCustomResource(obj)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	c, err := r.getScaleSetsClient(ctx)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleting the legacy VMSS %s", key.LegacyWorkerVMSSName(cr))) // nolint: errcheck
+	_, err = c.Delete(ctx, key.ResourceGroupName(cr), key.LegacyWorkerVMSSName(cr))
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleted the legacy VMSS %s", key.LegacyWorkerVMSSName(cr))) // nolint: errcheck
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleting the legacy VMSS deployment %s", legacyVMSSDeploymentName)) // nolint: errcheck
+
+	dc, err := r.getDeploymentsClient(ctx)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	_, err = dc.Delete(ctx, key.ResourceGroupName(cr), legacyVMSSDeploymentName)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleted the legacy VMSS deployment %s", legacyVMSSDeploymentName)) // nolint: errcheck
+
+	return DeploymentCompleted, nil
```

Comment:
Is it ok here that we don't wait for the VMSS and the deployment to actually be deleted?

When you call `c.Delete(ctx, key.ResourceGroupName(cr), key.LegacyWorkerVMSSName(cr))` and `dc.Delete(ctx, key.ResourceGroupName(cr), legacyVMSSDeploymentName)` above to delete legacy VMSS and its deployment, they are not still deleted, the functions return Futures.

So if you're not waiting for the Future to complete, then maybe:
- check if the VMSS exist, and if it exist and it's running (or just not being deleted), delete it; check if the deployment is there, and if it exist and not being deleted, delete it; return same state
- if both VMSS and deployment are deleted, set the state to `DeploymentCompleted`

Or just wait for the Futures to complete with something like this:
```golang
vmssDeleteFuture, err := c.Delete(ctx, key.ResourceGroupName(cr), key.LegacyWorkerVMSSName(cr))
if err != nil {
	return "", microerror.Mask(err)
}

_, err = c.DeleteResponder(vmssDeleteFuture.Response())
if err != nil {
	return "", microerror.Mask(err)
}
```

Reply:
We can't wait because the reconciliation loop would be too long.
You are right, we should check if the VMSS and deployments gets delete correctly.
What will happen if they don't? The VMSS simply stays there and we can get rid of it manually.
I know this doesn't sound like very "SRE", but adding new checks and complicating the code even more is not worth it IMHO.

Reply:
(late post-merge and post-vacation comment :smile:  )

> but adding new checks and complicating the code even more is not worth it IMHO

Agreed for this, as it was a one time thing for upgrading to Flatcar.